text/0000-cargo-report-future-incompat.md

```diff
@@ -0,0 +1,720 @@
+schedule established for when they are meant to become hard errors.
+(This RFC does not specify the details about how such schedules are
+established or what constraints they will have to meet; it just posits
+that they *will* be established by some means.) The metadata for every
+future-incompatibility lint should include the anticipated version of
+Rust, if known, where it will become a hard error.
+
+ * This adds motivation for using JSON formatted diagnostics: JSON
+   records are more readily extensible, and thus can support adding
+   this sort of feedback in a robust fashion.
+
+Since `cargo` is expected to continue to emit the report even when the
+upstream dependencies are not rebuilt, Cargo will store the
+future-incompatibility status for each crate somewhere in the `build/`
+directory on the file-system. (This is analogous to how incremental
+compilation caches diagnostic output, so that future runs will still
+emit the same report even when we do not recompile the same input.)
+
+
+## Policy issues
+
+We probably do not want to blindly convert every lint that is
+currently categorized as `C-future-incompatility` on the Rust repo to
+use this system. For example, there may be cases where we would be better
+served to use the edition mechanism to introduce the hard-error, but leave
+the functionality unchanged for older Rust editions. In any case, that
+
+is a policy matter for the relevant teams, and is out of scope for
+this RFC. I just want to make it clear that the mechanism suggested
+here may not be appropriate for every single lint currently
+categorized as `C-future-incompatility`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The change as described requires revisions to both `rustc` and
+`cargo`, which can be tricky to coordinate.
+
+This RFC suggests an approach where the changes are somewhat loosely
+coupled: Use of `--error-format=json` will enable the
+future-compatibility checking mode. This avoids the need to add a new
+stable command line flag to `rustc`; but it also may be a confusing
+change in compiler semantics for any non-cargo client of `rustc` that
+is using `--error-format=json`.
+
+Emitting a warning on all subsequent cargo invocations until the
+problem is resolved may be overkill. In particular, it may not be
+reasonable for someone to resolve this problem in the short term.
+
+ * (Perhaps we could restrict it so that it only gets re-emitted once
+   each day or some other way to modulate the annoyance level; but of
+   course then the diagnostic would have to state very clearly that it
+   is doing that strange magic.)
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## No change needed?
+
+Some claim that "our current approach" has been working, and therefore
+no change is needed here. However, my counterargument is that the only
+reason we haven't had to resolve this before is that compiler and
+language teams have been very conservative in changing
```

Comment:
I don't think this is true. Historically, I think not much has happened to C-future-compatibility lints because no one has been keen on transitioning them (but still, we have, e.g. https://github.com/rust-lang/rust/pull/63247 and the NLL stuff...) as that requires a non-trivial amount of work and because cap-lints has been in the way.

Reply:
.... I don't understand how what you wrote is supporting your statement "I don't think this is true": what you wrote seems, if anything, to agree that the teams *have* been conservative in their dealing with C-future-compatibility lints.

Is your problem with the sentence just a matter of phrasing, i.e. perhaps the current wording somehow implies that the teams *desired* to be conservative, which in fact the teams have been **forced** to be conservative (by constraints such a cap-lints and the associated amount of work) ?

Reply:
> Is your problem with the sentence just a matter of phrasing, i.e. perhaps the current wording somehow implies that the teams _desired_ to be conservative, which in fact the teams have been **forced** to be conservative (by constraints such a cap-lints and the associated amount of work) ?

Yes, that's it. I don't think we've desired as a matter of policy to be conservative.

Reply:
resolved by 8f4583f9fadc9dae868c161eec6288ebe4880ad4